crates/ty_python_semantic/src/types/narrow.rs

```diff
@@ -269,48 +271,244 @@ impl ClassInfoConstraintFunction {
     }
 }
 
+/// Represents a single conjunction (AND) of constraints in Disjunctive Normal
+/// Form (DNF).
+///
+/// A conjunction may contain: - A regular constraint (intersection of types) -
+/// An optional `TypeGuard` constraint that "replaces" the type rather than
+/// intersecting
+///
+/// For example, `(Conjunction { constraint: A, typeguard: Some(B) } &
+/// Conjunction { constraint: C, typeguard: Some(D)})` evlaluates to
+/// `Conjunction { constraint: C, typeguard: Some(D) }` because the type guard
+/// in the second clobbers the first.
+#[derive(Hash, PartialEq, Debug, Eq, Clone, Copy)]
+struct Conjunction<'db> {
+    /// The intersected constraints (represented as an intersection type)
+    constraint: Type<'db>,
+    /// If any constraint in this conjunction is a `TypeGuard`, this is Some and
+    /// contains the union of all `TypeGuard` types in this conjunction
+    typeguard: Option<Type<'db>>,
+}
+
+impl get_size2::GetSize for Conjunction<'_> {}
+
+impl<'db> Conjunction<'db> {
+    /// Create a new conjunction with just a regular constraint
+    fn regular(constraint: Type<'db>) -> Self {
+        Self {
+            constraint,
+            typeguard: None,
+        }
+    }
+
+    /// Create a new conjunction with a `TypeGuard` constraint
+    fn typeguard(constraint: Type<'db>) -> Self {
+        Self {
+            constraint: Type::object(),
+            typeguard: Some(constraint),
+        }
+    }
+
+    /// Evaluate this conjunction to a single type.
+    /// If there's a `TypeGuard` constraint, it replaces the regular constraint.
+    /// Otherwise, returns the regular constraint.
+    fn evaluate_type_constraint(self, db: &'db dyn Db) -> Type<'db> {
+        self.typeguard.map_or_else(
+            || self.constraint,
+            |typeguard_constraint| {
+                IntersectionBuilder::new(db)
+                    .add_positive(typeguard_constraint)
+                    .add_positive(self.constraint)
+                    .build()
+            },
+        )
+    }
+}
+
+/// Represents narrowing constraints in Disjunctive Normal Form (DNF).
+///
+/// This is a disjunction (OR) of conjunctions (AND) of constraints.
+/// The DNF representation allows us to properly track `TypeGuard` constraints
+/// through boolean operations.
+///
+/// For example:
+/// - `f(x) and g(x)` where f returns `TypeIs[A]` and g returns `TypeGuard[B]`
+///   => `[Conjunction { constraint: A, typeguard: Some(B) }]`
+///   => evaluates to `B` (`TypeGuard` clobbers)
+///
+/// - `f(x) or g(x)` where f returns `TypeIs[A]` and g returns `TypeGuard[B]`
+///   => `[Conjunction { constraint: A, typeguard: None }, Conjunction { constraint: object, typeguard: Some(B) }]`
+///   => evaluates to `A | B`
```

Comment:
Might be clearer here to say something like
```suggestion
///   => evaluates to `(P & A) | B` where `P` is our previously-known type
```